### PR TITLE
Add dev container feature

### PR DESCRIPTION
**Pull request by stuartleeks**
# Description

Add a `dev_container_feature` folder with a dev container feature that installs the Dapr CLI.
This can be used in any tooling that uses dev containers (VS Code, GitHub Codespaces) to enable developers to quickly and easily add the Dapr CLI to their project:

```jsonc
// Adding the following to your devcontainer.json will install the Dapr CLI
"features": {
    "ghcr.io/dapr/cli/dapr-cli:0": {}
}
```

Along with the feature code, the PR adds GitHub workflows for testing and releasing the dev container feature (packaged as OCI artifacts in GitHub container registry). This uses the pattern set out in the https://github.com/devcontainers/feature-starter repo, modified to handle the feature code being in a subfolder rather than the repo root.

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #1222 

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [X] Code compiles correctly
* [X] Created/updated tests
* [X] Extended the documentation

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -212,6 +212,20 @@ $ dapr init --container-runtime podman
 
 > Note: The default container runtime is Docker.
 
+#### In a dev container or GitHub Codespace
+
+To install Dapr in a dev container or GitHub Codespace, add the following to your `devcontainer.json` file:
```
Comment (ItalyPaleAle) [inline README.md:217]:
```suggestion
To install the Dapr CLI in a dev container or GitHub Codespace, add the following to your `devcontainer.json` file:
```

File: README.md [hunk 2]
```diff
@@ -212,6 +212,20 @@ $ dapr init --container-runtime podman
...
+
+```jsonc
+"features": {
+    "ghcr.io/dapr/cli/dapr-cli:0": {}
+}
+```
+
+Restart your dev container or GitHub Codespace and run `dapr init` to initialise the Dapr CLI.
```
Comment (ItalyPaleAle) [inline README.md:225]:
```suggestion
Restart your dev container or GitHub Codespace and run `dapr init` to initialize the Dapr CLI.
```

(We normally use the American spelling :) )

File: dev-container-feature/src/dapr-cli/install.sh [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  grep '"tag_name":' | sed -E "s/.*\"v([^\"]+)\".*/\1/"
+}
+
+VERSION=${VERSION:-"latest"}
+
+if [ "${VERSION}" = "latest" ]; then
+  VERSION=$(get_latest_release)
+fi
+
+curl -SsL https://github.com/dapr/cli/releases/download/v"${VERSION}"/dapr_linux_amd64.tar.gz | \
```
Comment (ItalyPaleAle) [inline dev-container-feature/src/dapr-cli/install.sh:17]: Please make sure to detect the system's architecture, as we want the feature to be usable on ARM64 (and maybe ARM 32-bit) too

File: dev-container-feature/src/dapr-cli/install.sh [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+dapr --version
+
+## Write bash completion code to a file and source if from .bash_profile
+mkdir -p $_REMOTE_USER_HOME/.dapr
+dapr completion bash >  $_REMOTE_USER_HOME/.dapr/completion.bash.inc
+printf "
+## dapr shell completion
+source '$_REMOTE_USER_HOME/.dapr/completion.bash.inc'
+" >> $_REMOTE_USER_HOME/.bashrc
```
Comment (ItalyPaleAle) [inline dev-container-feature/src/dapr-cli/install.sh:28]: Should we add this to the `.zshrc` too (if it exists)?

File: dev-container-feature/src/dapr-cli/devcontainer-feature.json [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+  "version": "0.1.0",
+  "description": "Install the Dapr CLI",
+  "options": {
+    "version": {
+      "type": "string",
+      "proposals": [
+        "latest"
+      ],
+      "default": "latest",
+      "description": "Enter a version (or latest)"
```
Comment (ItalyPaleAle) [inline dev-container-feature/src/dapr-cli/devcontainer-feature.json:13]:
```suggestion
      "description": "Version of the Dapr CLI to install (or \"latest\")"
```

File: dev-container-feature/src/dapr-cli/README.md [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+"features": {
+    "ghcr.io/dapr/cli/dapr-cli:0": {}
+}
+```
+
+## Options
+
+| Options Id | Description | Type | Default Value |
+|-----|-----|-----|-----|
+| version | Enter a version (or latest) | string | latest |
```
Comment (ItalyPaleAle) [inline dev-container-feature/src/dapr-cli/README.md:18]:
```suggestion
| version | Version of the Dapr CLI to install (or "latest") | string | latest |
```

**Review by ItalyPaleAle — CHANGES_REQUESTED** [review]
Thanks for this, @stuartleeks, we actually were talking about this for a few weeks (CC: @paulyuk ). Just a few comments on the code but overall it's very good

File: dev-container-feature/src/dapr-cli/install.sh (hunk 3 above)
Comment (stuartleeks) [reply dev-container-feature/src/dapr-cli/install.sh]: Good point - added :-)

File: .github/workflows/dev_container_feature_test.yaml [hunk 7]
```diff
@@ -0,0 +1,62 @@
+name: "CI - Test Features"
+on:
+  push:
+    branches:
+      - main
```
Comment (pravinpushkar) [inline .github/workflows/dev_container_feature_test.yaml:5]:
```suggestion
      - master
```

File: README.md [hunk 8]
```diff
@@ -212,6 +212,20 @@ $ dapr init --container-runtime podman
...
+
+To install the Dapr CLI in a dev container or GitHub Codespace, add the following to your `devcontainer.json` file:
+
+```jsonc
+"features": {
+    "ghcr.io/dapr/cli/dapr-cli:0": {}
+}
+```
+
+Restart your dev container or GitHub Codespace and run `dapr init` to initialize the Dapr CLI.
```
Comment (pravinpushkar) [inline README.md:225]:
```suggestion
Restart your dev container or GitHub Codespace and run `dapr init` to initialize Dapr.
```

File: .github/workflows/dev_container_feature_release.yaml [hunk 9]
```diff
@@ -0,0 +1,47 @@
+name: "Release dev container features & Generate Documentation"
+on:
+  workflow_dispatch:
+
+jobs:
+  deploy:
+    if: ${{ github.ref == 'refs/heads/main' }}
```
[the remark this answers is not in the document]
Comment (stuartleeks) [reply .github/workflows/dev_container_feature_release.yaml]: 👍

File: .github/workflows/dev_container_feature_test.yaml (hunk 7 above)
Comment (stuartleeks) [reply .github/workflows/dev_container_feature_test.yaml]: 👍

File: dev-container-feature/src/dapr-cli/install.sh (hunk 4 above)
Comment (stuartleeks) [reply dev-container-feature/src/dapr-cli/install.sh:63]: I tried adding to `.zshrc` but I don't get tab completion in zsh. I also tried running `source <(dapr completion zsh)` as per the output from `dapr completion` but still don't get completions.

I'm not really familiar with zsh - how big a deal is this for getting the initial PR merged?

Comment (ItalyPaleAle) [reply dev-container-feature/src/dapr-cli/install.sh:63]: That's ok, it's just for shell completion anyways.

File: .devcontainer/devcontainer.json [hunk 10]
```diff
@@ -58,6 +59,7 @@
 		// Run the entrypoint defined in container image.
 		"--init"
 	],
+    "postCreateCommand": "npm install -g @devcontainers/cli",
```
Comment (ItalyPaleAle) [inline .devcontainer/devcontainer.json:62]: Totally a nit, but this is indented using spaces and the rest uses tabs

**Review by ItalyPaleAle — DISMISSED** [review]

File: .devcontainer/devcontainer.json (hunk 10 above)
Comment (stuartleeks) [reply .devcontainer/devcontainer.json]: Good spot - fixed. While I was there I also tweaked the config to remove the schema validation warnings in `devcontainer.json` 😄

**Review by ItalyPaleAle — APPROVED** [review]
LGTM :) Thanks

**Comment by pravinpushkar** [comment]
vulnerability check failure is addressed here- https://github.com/dapr/cli/pull/1237

**Comment by mukundansundar** [comment]
@stuartleeks Quick question regarding the devcontainer feature ... 
So For releasing this as a feature, we need to manually run the workflow ?

**Comment by mukundansundar** [comment]
> lly run the feature release workflow

I will finish the review by tomorrow ... Thanks again for this contribution!

**Review by pravinpushkar — APPROVED** [review]
lgtm

**Review by shubham1172 — APPROVED** [review]

File: .github/workflows/dev_container_feature_release.yaml [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+      - name: Create PR for Documentation
+        id: push_image_info
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        run: |
+          set -e
+          echo "Start."
+          # Configure git and Push updates
+          git config --global user.email github-actions@github.com
+          git config --global user.name github-actions
```
Comment (mukundansundar) [inline .github/workflows/dev_container_feature_release.yaml:35]: Shouldn't this be related to daprbot?
@artursouza @yaron2 @ItalyPaleAle  thoughts?

Comment (ItalyPaleAle) [reply .github/workflows/dev_container_feature_release.yaml:35]: This is ok with me, but I am also ok with changing to use Dapr bot. Unsure if these values are used as convention and there’s any benefit.

Comment (stuartleeks) [reply .github/workflows/dev_container_feature_release.yaml:35]: If you have different values that you'd like used for this I'm happy to push an update - let me know 😄

Comment (artursouza) [reply .github/workflows/dev_container_feature_release.yaml:35]: It is OK to be via github-actions.

**Review by mukundansundar — APPROVED** [review]

**Comment by mukundansundar** [comment]
the go vulnerability check should ideally not fail if it uses 1.20.2 version of go but sadly the rerun also used only 1.20.1 version of go .. Will try to see what can be done about it

**Comment by stuartleeks** [comment]
@mukundansundar - From what I can see, `actions/setup-go` is reading `go.mod` which specifies `1.20` as the version to install. However, it looks like it is finding `1.20.1` in the tools cache on the agent (which matches the [current runner docs](https://github.com/actions/runner-images/blob/77039ebf7edb29fc43862359b3ddf2b2b6e01c09/images/linux/Ubuntu2204-Readme.md#go))

**Comment by mukundansundar** [comment]
Exactly... 
Either go vulnerability check has to be made manual and disabled in the workflow or go version has to be made manual and I do not want to specify go version with the patch version every time. 
So thinking of the first option. If the vulnerability check is made manual with a make command, it can be run periodically to check for vulnerabilities and fixed as well as before release....

**Comment by stuartleeks** [comment]
@mukundansundar it looks like the hosted runners should have `1.20.2` in a couple of days: https://github.com/actions/runner-images/issues/7281#issuecomment-1468281955

**Comment by stuartleeks** [comment]
@mukundansundar thanks for you work to get this merged!

Is it possible to run the [dev_container_feature_release.yaml](https://github.com/dapr/cli/actions/workflows/dev_container_feature_release.yaml) workflow to package and publish the dev container feature?

**Comment by stuartleeks** [comment]
@ItalyPaleAle - thanks for PR to update the docs, I'll submit a PR here to update the README.

I've also [raised a PR ](https://github.com/devcontainers/devcontainers.github.io/compare/gh-pages...stuartleeks:devcontainers.github.io:patch-1)to have the feature included in the index (so that it shows up in the feature list/search in VS Code)

**Pull request merged**